src/editor/Model/WindowManager.re

```diff
@@ -0,0 +1,115 @@
+open Revery_UI;
+
+module WindowSplitId =
+  Revery.UniqueId.Make({});
+
+[@deriving show]
+type componentCreator = unit => React.syntheticElement;
+
+[@deriving show]
+type direction =
+  | Horizontal
+  | Vertical;
+
+[@deriving show]
+type split = {
+  id: int,
+  parentId: int,
+  component: componentCreator,
+  direction,
+  /* if omitted the split will grow to occupy whatever space is available */
+  width: option(int),
+  height: option(int),
+};
+
+[@deriving show]
+type dock = {
+  component: componentCreator,
+  width: option(int),
+  height: option(int),
+};
+
+[@deriving show]
+type splitTree =
+  | Parent(direction, int, list(splitTree))
+  | Leaf(split);
+
+[@deriving show]
+type splits = splitTree;
+
+type t = {
+  windows: splits,
+  leftDock: option(dock),
+  rightDock: option(dock),
+};
+
+let empty = Parent(Vertical, 0, []);
+
+let create = (): t => {windows: empty, leftDock: None, rightDock: None};
+
+let getId = (id: option(int)) =>
+  switch (id) {
+  | Some(i) => i
+  | None => WindowSplitId.getUniqueId()
+  };
+
+let createSplit =
+    (~id=?, ~parentId, ~width=?, ~height=?, ~component, ~direction, ()) => {
+  id: getId(id),
+  parentId,
+  component,
+  width,
+  height,
+  direction,
+};
+
+let createDock = (~width=?, ~height=?, ~component, ()) => {
+  width,
+  height,
+  component,
+};
+
+type parentHandler('a) = ('a, direction) => 'a;
+
+let identity: parentHandler('a) = (result: 'a, _direction) => result;
+
+let rec traverseSplitTree =
+        (
+          ~handleParent=identity,
+          ~action,
+          ~result,
+          ~tree,
+          ~direction: direction,
+          (),
+        ) =>
+  switch (tree) {
+  | Parent(direction, _, children) =>
+    handleParent(result, direction)
+    |> (
+      newResult =>
+        List.fold_left(
+          (accum, child) =>
+            traverseSplitTree(
+              ~action,
+              ~result=accum,
+              ~tree=child,
+              ~handleParent,
+              ~direction,
+              (),
+            ),
+          newResult,
+          children,
+        )
+    )
+  | Leaf(split) => action(result, split, direction)
+  };
+
+let rec add = (id, split, tree) =>
+  switch (tree) {
+  | Parent(direction, parentId, children) when id == parentId =>
```

Comment:
Do we have to special case if, for example, the Parent has a split direction of `Horizontal`, but we request a `Vertical` split with that parent id? 🤔

Reply:
that and the `remove` function are still a work in progress, I haven't written down any interim thoughts but was thinking if the new split is changing direction then we'd create a new parent and place it inside but I need to experiment more with the idea. I'll see what I come up with

Reply:
Ah ok! That plan to add a new parent inside if the direction is different sounds like a good approach to me.